### PR TITLE
feat: plugin to remove attachments from posts

### PR DESCRIPTION
**Pull request by fmartingr**
#### Summary
- Adds a post menu option to remove attachments
- Only post author or user with permission to edit other's post can remove attachments
- Respects the `PostEditTimeLimit` setting

#### Ticket link
https://mattermost.atlassian.net/browse/MM-59951

### CODE REVIEW

File: server/api.go
Comment (wiggin77) [inline server/api.go]: The customer has the setting enabled that prevents users from deleting posts.  
image

They also do not allow editing posts after minutes.  I think we should honour those settings.

Comment (wiggin77) [reply server/api.go]: Not sure if `userHasRemovePermissionsToPost` or `DeletePost` respect those settings already.

Comment (fmartingr) [reply server/api.go]: it honors the edit time, not the delete post. It would be weird because a post with no message will just be a name with nothing. Maybe I can put a message like `(attachments removed)` or something like that for empty messages?

Comment (fmartingr) [reply server/api.go]: Ok I have removed the code block that delete posts entirely, rather than checking the permission as well to avoid more complexity.

File: server/api.go
Comment (wiggin77) [inline server/api.go:53]: Typo

Comment (wiggin77) [inline server/api.go]: Typo

File: server/sqlstore/file.go
Comment (wiggin77) [inline server/sqlstore/file.go]: Typo

**Review by wiggin77 — CHANGES_REQUESTED** [review]
Some typos to fix.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:   Great work.

**Comment by wiggin77** [comment]
@hanzei @gabrieljackson this PR is time sensitive. We plan to have some QA done once merged before shipping to customer end of month.

**Comment by wiggin77** [comment]
@fmartingr once this is merged, Saturn has agreed to have his team do some QA.  Please assign to him when ready.

**Comment by hanzei** [comment]
@fmartingr @wiggin77 Sorry, I don't have the bandwidth to review the PR this week. Can you please ask someone else for a review? Thanks!

File: server/api.go
Comment (esarafianou) [inline server/api.go]: We need to specify that it's a `DELETE` HTTP method. Right now, we're deleting data with GET.

File: server/permissions.go
Comment (esarafianou) [inline server/permissions.go]: We need to check if the user can edit their own posts. If they are not allowed to edit their posts, they shouldn't be allowed to delete the attachments of a post. 

here's how the check is in `patchPost` in the server:

```

	var permission *model.Permission
	if c.AppContext.Session().UserId == originalPost.UserId {
		permission = model.PermissionEditPost
	} else {
		permission = model.PermissionEditOthersPosts
	}

	if !c.App.SessionHasPermissionToChannel(c.AppContext, *c.AppContext.Session(), originalPost.ChannelId, permission) {
		c.SetPermissionError(permission)
		return
	}
```

ref: https://github.com/mattermost/mattermost/blob/master/server/channels/api4/post.go#L957-L967

File: server/api.go
Comment (esarafianou) [inline server/api.go:71]: Files are still visible on the RHS of the channel and I think it's because we attach them to the original post ID that still belongs to the channel. 
0/5 if this is an issue or not. It depends on what's the product goal of the plugin.

Screenshot 2024-08-22 at 12 29 20

**Review by esarafianou — CHANGES_REQUESTED** [review]

**Comment by esarafianou** [comment]
@fmartingr one more comment. This bullet from the description doesn't hold true based on the implementation and my testing. 

> If post contains no message (only attachments) the post is removed

Edit: I see in a comment thread with Doug that you removed this functionality. Let's update the description to make this clear.

File: server/api.go
Comment (esarafianou) [reply server/api.go]: Suggestion to make it RESTful:
From:
```
GET /plugins/com.mattermost.attachments-remover/api/v1/remove_attachments?post_id=<POST ID>
```
to 
```
DELETE  /plugins/com.mattermost.attachments-remover/api/v1/posts/{post_id}/attachments
```

If you still want the post ID as query parameter, we could also go with `DELETE /plugins/com.mattermost.attachments-remover/api/v1/attachments?post_id=<POST ID>` although that's less RESTful.

Comment (fmartingr) [reply server/api.go]: I need to check but request should be a POST request.

I'm using a Mattemost [interactive dialog](https://developers.mattermost.com/integrate/plugins/interactive-dialogs/) to easily display the confirmation modal and I can't customize the HTTP verb in there.

File: server/api.go
Comment (fmartingr) [reply server/api.go:71]: Ummmm, something should have changed, this one of the first things I tested myself. Let me check what's going on.

Comment (fmartingr) [reply server/api.go:71]: Ok the problem is the query that removes the channelid from the file, is not updating the row nor giving an error. Investigating.

File: server/api.go
Comment (fmartingr) [reply server/api.go]: https://github.com/mattermost/mattermost-plugin-attachments-remover/pull/1/commits/9b9c67b49bdb50ca3f87b078c33c5435c75a8cea

File: server/permissions.go
Comment (fmartingr) [reply server/permissions.go]: https://github.com/mattermost/mattermost-plugin-attachments-remover/pull/1/commits/92c1aa20b7e6f500a44edd070de0c70f5b180e6a

File: server/api.go
Comment (fmartingr) [reply server/api.go:71]: https://github.com/mattermost/mattermost-plugin-attachments-remover/pull/1/commits/a6c4f06ddf2b26af63783e505a098a59187db94f

**Review by esarafianou — APPROVED** [review]
Thanks @fmartingr!

File: server/api.go
Comment (gabrieljackson) [inline server/api.go:68]: Just checking that we wouldn't want to proceed on error here with just logging it. 0/5 on that being better, but just wanted to check. When doing Wrangler stuff I sometimes opted to try to finish a given task once a "breaking change" was made and just log errors. Once again, just checking if that is something we thought about for unhappy paths.

**Review by gabrieljackson — APPROVED** [review]
Looks good! Just one non-blocking question.

File: server/api.go
Comment (wiggin77) [reply server/api.go:68]: @fmartingr I'll let you decide if you want do this via a new PR.  A new PR could make best efforts to delete each attachment, continuing on error and logging.  The attaching IDs to original post would then work off a new slice built in this loop.

**Review by wiggin77 — APPROVED** [review]

File: server/api.go
Comment (fmartingr) [reply server/api.go:68]: The thing is that removing attachments one by one will create a new post each time I remove an attachment if I do it this way (since I'm using `UpdatePost`) and seems a bit dirty. Something like this:

```go
	for _, fileID := range post.FileIds {
		post.FileIds = slices.DeleteFunc(post.FileIds, func(e string) bool {
			return e == fileID
		})
		updatedPost, appErr := a.plugin.API.UpdatePost(post)
		if appErr != nil {
			a.plugin.API.LogError("error updating post to remove attachment", "err", appErr, "post_id", postID, "file_id", fileID)
			continue
		}

		if err := a.plugin.SQLStore.DetachAttachmentFromChannel(fileID); err != nil {
			a.plugin.API.LogError("error detaching attachment from channel", "err", err, "post_id", postID, "file_id", fileID)
			continue
		}

		// Attach the original file IDs to the original post so history is not lost
		if err := a.plugin.SQLStore.AttachFileIDsToPost(updatedPost.OriginalId, []string{fileID}); err != nil {
			a.plugin.API.LogError("error attaching original file IDs to post", "err", err, "post_id", postID, "file_id", fileID)
			continue
		}
	}
```

And then we would need to discuss... should I make it atomic (that is, if last step fails should I revert the ones before)?

The easier way seems to be:
```suggestion
	for _, fileID := range originalFileIDs {
		if err := a.plugin.SQLStore.DetachAttachmentFromChannel(fileID); err != nil {
			a.plugin.API.LogError("error detaching attachment from channel", "err", err", post_id", postID, "file_id", fileID)
		}
	}
```

**Pull request merged**
